integ/src/nodegroup_provider.rs

```diff
@@ -0,0 +1,573 @@
+        .instance_profile_name(&iam_instance_profile_name.clone())
+        .send()
+        .await
+        .context("Unable to create instance profile")?;
+    iam_client
+        .delete_role()
+        .role_name(&iam_instance_profile_name.clone())
+        .send()
+        .await
+        .context("Unable to delete role.")?;
+
+    Ok(())
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=  =^..^=  Related sub-functions of sources creation and termination   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+// Find the node ami id to use.
+async fn find_ami_id(
+    ssm_client: &aws_sdk_ssm::Client,
+    arch: &str,
+    br_version: &str,
+    eks_version: &str,
+) -> ProviderResult<String> {
+    let parameter_name = format!(
+        "/aws/service/bottlerocket/aws-k8s-{}/{}/{}/image_id",
+        eks_version, arch, br_version
+    );
+    let ami_id = ssm_client
+        .get_parameter()
+        .name(parameter_name)
+        .send()
+        .await
+        .context("Unable to get ami id")?
+        .parameter
+        .context("Unable to get ami id")?
+        .value
+        .context("ami id is missing")?;
+    Ok(ami_id)
+}
+
+/// Determine the instance type to use. If provided use that one. Otherwise, for `x86_64` use `m5.large`
+/// and for `aarch64` use `m6g.large`
+async fn instance_type(ec2_client: &aws_sdk_ec2::Client, node_ami: &str) -> ProviderResult<String> {
+    let arch = ec2_client
+        .describe_images()
+        .image_ids(node_ami)
+        .send()
+        .await
+        .context("Unable to get ami architecture")?
+        .images
+        .context("Unable to get ami architecture")?
+        .get(0)
+        .context("Unable to get ami architecture")?
+        .architecture
+        .clone()
+        .context("Ami has no architecture")?;
+
+    Ok(match arch {
+        ArchitectureValues::X8664 => "m5.large",
+        ArchitectureValues::Arm64 => "m6g.large",
+        _ => "m6g.large",
+    }
+    .to_string())
+}
+
+fn first_subnet_id(subnet_ids: &[String]) -> ProviderResult<String> {
+    subnet_ids
+        .get(0)
+        .map(|id| id.to_string())
+        .context("There are no private subnet ids")
+}
+
+fn tag_specifications(cluster_name: &str) -> LaunchTemplateTagSpecificationRequest {
+    LaunchTemplateTagSpecificationRequest::builder()
+        .resource_type(ResourceType::Instance)
+        .tags(
+            Tag::builder()
+                .key("Name")
+                .value(format!("{}_node", cluster_name))
+                .build(),
+        )
+        .tags(
+            Tag::builder()
+                .key(format!("kubernetes.io/cluster/{}", cluster_name))
+                .value("owned")
+                .build(),
+        )
+        .tags(
+            Tag::builder()
+                .key(INSTANCE_TAG_NAME)
+                .value(INSTANCE_TAG_VALUE)
+                .build(),
+        )
+        .build()
+}
+
+fn userdata(endpoint: &str, cluster_name: &str, certificate: &str) -> String {
```

Comment:
We should probably use a [TOML serializer](https://docs.rs/toml/latest/toml/ser/fn.to_string_pretty.html) with serde for this, since it would be pretty easy to break the TOML formatting with the right input.

Reply:
Do you mean something like below? I tried this but `create_launch_template().user_data( input)`, the input need a valid base64 string. Can I still use TOML serializer to output base64 string? 
```
n userdata(endpoint: &str, cluster_name: &str, certificate: &str) -> ProviderResult<String> {
    let user_data = toml::ser::to_string_pretty(&format!(
        r#"[settings.updates]
            ignore-waves = true
                
            [settings.kubernetes]
            api-server = "{}"
            cluster-name = "{}"
            cluster-certificate = "{}"
        "#,
        endpoint, cluster_name, certificate
    ))
    .context("Unable convert userdata to pretty string")?;

    Ok(user_data)
```

Reply:
I meant more like create a series of structs like:

```rust
#[derive(Serialize)]
struct Settings {
    updates: UpdateSettings,
    kubernetes: KubernetesSettings,
}

#[derive(Serialize)]
struct UpdateSettings {
    ignore_waves: bool,
}

#[derive(Serialize)]
struct KubernetesSettings {
    api_server: String,
    // etc...
}
```

Then you could do something like:

```rust
let settings = Settings {
    updates: UpdateSettings {
        ignore_waves: true,
    },
    kubernetes: KubernetesSettings {
    },
}
toml::ser::to_string_pretty(settings);
```

Reply:
I tried the method that @cbgbt mentioned above; however, the string formate of userdata which was created by `toml::ser::to_string_pretty()` can be recognized by 'early-boot-config'. We can't use this method and I think it's better to keep the original method.
Error
```
[    5.826817] early-boot-config[1483]: Provider error: Unable to serialize settings from instance user data: TOML data did not contain 'settings' section
```
userdata created by toml
```
"[\"settings.updates\"]\n
ignore-waves = true\n\n
[\"settings.kubernetes\"]\n
api-server = 'xxxx7.us-west-2.eks.amazonaws.com'\n
cluster-name = 'bottlerocket-test'\n
cluster-certificate = '=xxxxx'\n"
```